File: src/main/python/net/coderextreme/data/geobubbles.py
print('<!--')
import x3d
print('-->')
X3D0 = x3d.X3D()
X3D0.profile = "Immersive"
X3D0.version = "4.1"
head1 = x3d.head()
component2 = x3d.component()
component2.name = "Geospatial"
component2.level = 1

head1.children.append(component2)
meta3 = x3d.meta()
meta3.name = "title"
meta3.content = "geobubbles.x3d"

head1.children.append(meta3)
meta4 = x3d.meta()
meta4.name = "creator"
meta4.content = "John Carlson"

head1.children.append(meta4)
meta5 = x3d.meta()
meta5.name = "generator"
meta5.content = "x3d-tidy V3.0.2, https://www.npmjs.com/package/x3d-tidy"

head1.children.append(meta5)
meta6 = x3d.meta()
meta6.name = "generator"
meta6.content = "manual"

head1.children.append(meta6)

X3D0.head = head1
Scene7 = x3d.Scene()
GeoViewpoint8 = x3d.GeoViewpoint(DEF="Tour")
GeoViewpoint8.description = "Tour Views"
GeoViewpoint8.position = [0.000213314646318662,0.00135748535376003,3.99999812897295]

Scene7.children.append(GeoViewpoint8)
Background9 = x3d.Background()
Background9.frontUrl = ["../resources/images/FR.png","https://coderextreme.net/X3DJSONLD/src/main/resources/images/FR.png"]
Background9.backUrl = ["../resources/images/BK.png","https://coderextreme.net/X3DJSONLD/src/main/resources/images/BK.png"]
Background9.leftUrl = ["../resources/images/LF.png","https://coderextreme.net/X3DJSONLD/src/main/resources/images/LF.png"]
Background9.rightUrl = ["../resources/images/RT.png","https://coderextreme.net/X3DJSONLD/src/main/resources/images/RT.png"]
Background9.topUrl = ["../resources/images/TP.png","https://coderextreme.net/X3DJSONLD/src/main/resources/images/TP.png"]
Background9.bottomUrl = ["../resources/images/BT.png","https://coderextreme.net/X3DJSONLD/src/main/resources/images/BT.png"]

Scene7.children.append(Background9)
Transform10 = x3d.Transform()
Shape11 = x3d.Shape()
Appearance12 = x3d.Appearance()
Material13 = x3d.Material()
Material13.diffuseColor = [0.7,0.7,0.7]
Material13.specularColor = [0.5,0.5,0.5]

Appearance12.material = Material13

Shape11.appearance = Appearance12
Sphere14 = x3d.Sphere()

Shape11.geometry = Sphere14

Transform10.children.append(Shape11)

Scene7.children.append(Transform10)
TimeSensor15 = x3d.TimeSensor(DEF="TourTime")
TimeSensor15.cycleInterval = 5
TimeSensor15.loop = True

Scene7.children.append(TimeSensor15)
GeoPositionInterpolator16 = x3d.GeoPositionInterpolator(DEF="TourPosition")
GeoPositionInterpolator16.key = [0,1]
GeoPositionInterpolator16.keyValue = [(0.0015708, 0, 4),(0, 0.0015708, 4)]

Scene7.children.append(GeoPositionInterpolator16)
Script17 = x3d.Script(DEF="RandomTourTime")
field18 = x3d.field()
field18.accessType = "inputOnly"
field18.type = "SFTime"
field18.name = "set_cycle"

Script17.field.append(field18)
field19 = x3d.field()
field19.accessType = "inputOutput"
field19.type = "SFFloat"
field19.name = "val"

Script17.field.append(field19)
field20 = x3d.field()
field20.accessType = "inputOutput"
field20.type = "MFVec3d"
field20.name = "positions"
field20.value = [(0.0015708, 0, 4),(0, 0.0015708, 4)]

Script17.field.append(field20)
field21 = x3d.field()
field21.accessType = "inputOutput"
field21.type = "MFVec3d"
field21.name = "position"
field21.value = [(0.0015708, 0, 4),(0, 0.0015708, 4)]

Script17.field.append(field21)

Script17.sourceCode = '''ecmascript:\n"+
"\n"+
"               function set_cycle(value) {\n"+
"                        var cartesianMult = -150;  // -150 if cartesian, 1 if geo\n"+
"                        var ov = val;\n"+
"			// Browser.print('old '+ov);\n"+
"                        do {\n"+
"                                val = Math.floor(Math.random()*2);\n"+
"                                var vc = val;\n"+
"                                positions[vc] = new SFVec3d(Math.round(Math.random()*2)*0.0015708*cartesianMult, Math.round(Math.random()*2)*0.0015708*cartesianMult, 4);\n"+
"                        } while ( positions[ov][0] === positions[vc][0] && positions[ov][1] === positions[vc][1] && positions[ov][2] === positions[vc][2]);\n"+
"			// Browser.println(positions[ov]);\n"+
"			// Browser.println(positions[vc]);\n"+
"                        position = new MFVec3d();\n"+
"                        position[0] = new SFVec3d(positions[ov][0],positions[ov][1],positions[ov][2]);\n"+
"                        position[1] = new SFVec3d(positions[vc][0],positions[vc][1],positions[vc][2]);\n"+
"               }'''

Scene7.children.append(Script17)
ROUTE22 = x3d.ROUTE()
ROUTE22.fromNode = "TourTime"
ROUTE22.fromField = "cycleTime"
ROUTE22.toNode = "RandomTourTime"
ROUTE22.toField = "set_cycle"

Scene7.children.append(ROUTE22)
ROUTE23 = x3d.ROUTE()
ROUTE23.fromNode = "RandomTourTime"
ROUTE23.fromField = "position_changed"
ROUTE23.toNode = "TourPosition"
ROUTE23.toField = "set_keyValue"

Scene7.children.append(ROUTE23)
ROUTE24 = x3d.ROUTE()
ROUTE24.fromNode = "TourTime"
ROUTE24.fromField = "fraction_changed"
ROUTE24.toNode = "TourPosition"
ROUTE24.toField = "set_fraction"

Scene7.children.append(ROUTE24)
ROUTE25 = x3d.ROUTE()
ROUTE25.fromNode = "TourPosition"
ROUTE25.fromField = "geovalue_changed"
ROUTE25.toNode = "Tour"
ROUTE25.toField = "set_position"

Scene7.children.append(ROUTE25)

X3D0.Scene = Scene7
f = open("../data/geobubbles.new.python.x3d", mode="w", encoding="utf-8")
f.write(X3D0.XML())
f.close()
f = open("../data/geobubbles.new.python.x3dv", mode="w", encoding="utf-8")
f.write(X3D0.VRML())
f.close()
f = open("../data/geobubbles.new.python.x3dj", mode="w", encoding="utf-8")
f.write(X3D0.JSON())
f.close()

File: src/main/go/net/coderextreme/data/geobubbles.go

package main

import (
	"encoding/xml"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"

	"gox3d/x3d" // Assumes your module is named x3d-go-tester

	// Imports for the CGo-based libxml2 wrapper
	"github.com/lestrrat-go/libxml2"
	"github.com/lestrrat-go/libxml2/xsd"
)

// ... (Helper functions remain the same) ...
func stringPtr(s string) *string { return &s }
func floatPtr(f float32) *float32 { return &f }
func doublePtr(d float64) *float64 { return &d }
func boolPtr(b bool) *bool       { return &b }
func int32Ptr(i int32) *int32    { return &i }

func downloadSchemaIfNotExists(url, filepath string) error {
	// ... (function is correct, no changes) ...
	if _, err := os.Stat(filepath); err == nil {
		fmt.Printf("Schema file '%s' already exists. Skipping download.\n", filepath)
		return nil
	}
	fmt.Printf("Downloading X3D schema from %s to %s...\n", url, filepath)
	resp, err := http.Get(url)
	if err != nil { return fmt.Errorf("failed to download schema: %w", err) }
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK { return fmt.Errorf("bad status: %s", resp.Status) }
	out, err := os.Create(filepath)
	if err != nil { return fmt.Errorf("failed to create schema file: %w", err) }
	defer out.Close()
	_, err = io.Copy(out, resp.Body)
	if err != nil { return fmt.Errorf("failed to save schema: %w", err) }
	fmt.Println("Schema downloaded successfully.")
	return nil
}

// Updated validation function to print detailed errors
func validateXMLWithSchema(xmlData []byte, schemaPath string) error {
	schema, err := xsd.ParseFromFile(schemaPath)
	if err != nil { return fmt.Errorf("failed to parse schema file %s: %w", schemaPath, err) }
	defer schema.Free()
	doc, err := libxml2.Parse(xmlData)
	if err != nil { return fmt.Errorf("failed to parse XML data: %w", err) }
	defer doc.Free()

	validationErr := schema.Validate(doc)
	if validationErr != nil {
		errorLog := ""
		if schemaErrors, ok := validationErr.(xsd.SchemaValidationError); ok {
			for i, e := range schemaErrors.Errors() {
				errorLog += fmt.Sprintf("Error %d: %s\n", i+1, e.Error())
			}
			return fmt.Errorf("schema validation failed with %d errors:\n%s", len(schemaErrors.Errors()), errorLog)
		}
		return fmt.Errorf("schema validation failed: %w", validationErr)
	}
	return nil
}

func main() {
	fmt.Println("--- Building and Testing an X3D Scene in Go ---")

	const schemaURL = "https://www.web3d.org/specifications/x3d-4.0.xsd"
	const schemaFilename = "x3d-4.0.xsd"
	if err := downloadSchemaIfNotExists(schemaURL, schemaFilename); err != nil {
		log.Fatalf("Could not prepare schema file: %v", err)
	}
    sceneRoot := &x3d.X3D{
        Profile: stringPtr("Immersive"),
        Version: stringPtr("4.0"),
        Head: &x3d.Head{
            Components: []*x3d.Component{
                &x3d.Component{
                    Name: stringPtr("Geospatial"),
                    Level: int32Ptr(1),
            },
        },
            Metas: []*x3d.Meta{
                &x3d.Meta{
                    Name: stringPtr("title"),
                    Content: stringPtr("geobubbles.x3d"),
            },
            &x3d.Meta{
                Name: stringPtr("creator"),
                Content: stringPtr("John Carlson"),
            },
            &x3d.Meta{
                Name: stringPtr("generator"),
                Content: stringPtr("manual"),
            },
            &x3d.Meta{
                Name: stringPtr("generator"),
                Content: stringPtr("x3d-tidy V2.1.21, https://www.npmjs.com/package/x3d-tidy"),
            },
            },
        },
        Scene: &x3d.Scene{
            Children: []x3d.X3DChildNode{
                &x3d.GeoViewpoint{
                    CoreX3DNode: x3d.CoreX3DNode{
                        DEF: stringPtr("Tour"),
                    },
                    Description: stringPtr("Tour Views"),
                    Position: &x3d.SFVec3d{0.000240332395536092, 0.00133046760455211, 3.99999793432653},
                },
                &x3d.Background{
                    FrontUrl: x3d.MFString{"../resources/images/FR.png", "https://coderextreme.net/X3DJSONLD/src/main/resources/images/FR.png"},
                    BackUrl: x3d.MFString{"../resources/images/BK.png", "https://coderextreme.net/X3DJSONLD/src/main/resources/images/BK.png"},
                    LeftUrl: x3d.MFString{"../resources/images/LF.png", "https://coderextreme.net/X3DJSONLD/src/main/resources/images/LF.png"},
                    RightUrl: x3d.MFString{"../resources/images/RT.png", "https://coderextreme.net/X3DJSONLD/src/main/resources/images/RT.png"},
                    TopUrl: x3d.MFString{"../resources/images/TP.png", "https://coderextreme.net/X3DJSONLD/src/main/resources/images/TP.png"},
                    BottomUrl: x3d.MFString{"../resources/images/BT.png", "https://coderextreme.net/X3DJSONLD/src/main/resources/images/BT.png"},
                },
                &x3d.Transform{
                    Children: []x3d.X3DNode{
                        &x3d.Shape{
                            Appearance: &x3d.Appearance{
                                Material: &x3d.Material{
                                    DiffuseColor: &x3d.SFColor{0.7, 0.7, 0.7},
                                    SpecularColor: &x3d.SFColor{0.5, 0.5, 0.5},
                                },
                            },
                            Geometry: &x3d.Sphere{
                            },
                        },
                    },
                },
                &x3d.TimeSensor{
                    CoreX3DNode: x3d.CoreX3DNode{
                        DEF: stringPtr("TourTime"),
                    },
                    CycleInterval: doublePtr(5.0),
                    Loop: boolPtr(true),
                },
                &x3d.GeoPositionInterpolator{
                    CoreX3DNode: x3d.CoreX3DNode{
                        DEF: stringPtr("TourPosition"),
                    },
                    Key: x3d.MFFloat{0.0, 1.0},
                    KeyValue: &x3d.MFVec3d{[3]float64{0.0015708,0,4},[3]float64{0,0.0015708,4}},
                },
                &x3d.Script{
                    CoreX3DNode: x3d.CoreX3DNode{
                        DEF: stringPtr("RandomTourTime"),
                    },
                    Field: []x3d.X3DNode{
                        &x3d.Field{
                            AccessType: stringPtr("inputOnly"),
                            Type: stringPtr("SFTime"),
                            Name: stringPtr("set_cycle"),
                    },
                    &x3d.Field{
                        AccessType: stringPtr("inputOutput"),
                        Type: stringPtr("SFFloat"),
                        Name: stringPtr("val"),
                    },
                    &x3d.Field{
                        AccessType: stringPtr("inputOutput"),
                        Type: stringPtr("MFVec3d"),
                        Name: stringPtr("positions"),
                        Value: stringPtr("0.0015708 0 4 0 0.0015708 4"),
                    },
                    &x3d.Field{
                        AccessType: stringPtr("inputOutput"),
                        Type: stringPtr("MFVec3d"),
                        Name: stringPtr("position"),
                        Value: stringPtr("0.0015708 0 4 0 0.0015708 4"),
                    },
//ecmascript:
//
//               function set_cycle(value) {
//                        var cartesianMult = -150;  // -150 if cartesian, 1 if geo
//                        var ov = val;
//			// Browser.print('old '+ov);
//                        do {
//                                val = Math.floor(Math.random()*2);
//                                var vc = val;
//                                positions[vc] = new SFVec3d(Math.round(Math.random()*2)*0.0015708*cartesianMult, Math.round(Math.random()*2)*0.0015708*cartesianMult, 4);
//                        } while ( positions[ov][0] === positions[vc][0] && positions[ov][1] === positions[vc][1] && positions[ov][2] === positions[vc][2]);
//			// Browser.println(positions[ov]);
//			// Browser.println(positions[vc]);
//                        position = new MFVec3d();
//                        position[0] = new SFVec3d(positions[ov][0],positions[ov][1],positions[ov][2]);
//                        position[1] = new SFVec3d(positions[vc][0],positions[vc][1],positions[vc][2]);
//               }
                    },
                },
                &x3d.ROUTE{
                    FromNode: stringPtr("TourTime"),
                    FromField: stringPtr("cycleTime"),
                    ToNode: stringPtr("RandomTourTime"),
                    ToField: stringPtr("set_cycle"),
                },
                &x3d.ROUTE{
                    FromNode: stringPtr("RandomTourTime"),
                    FromField: stringPtr("position_changed"),
                    ToNode: stringPtr("TourPosition"),
                    ToField: stringPtr("set_keyValue"),
                },
                &x3d.ROUTE{
                    FromNode: stringPtr("TourTime"),
                    FromField: stringPtr("fraction_changed"),
                    ToNode: stringPtr("TourPosition"),
                    ToField: stringPtr("set_fraction"),
                },
                &x3d.ROUTE{
                    FromNode: stringPtr("TourPosition"),
                    FromField: stringPtr("geovalue_changed"),
                    ToNode: stringPtr("Tour"),
                    ToField: stringPtr("set_position"),
                },
            },
        },
    }


	fmt.Println("\n--- Validating the generated scene (internal logic) ---")
	err := sceneRoot.Scene.Validate()
	if err != nil {
		log.Fatalf("Scene validation failed: %v", err)
	}
	fmt.Println("✅ Internal scene validation successful!")
	fmt.Println("\n--- Serializing scene to XML ---")
	output, err := xml.MarshalIndent(sceneRoot, "", "  ")

	if err != nil {
		log.Fatalf("XML Marshaling failed: %v", err)
	}
	/*
	fmt.Println("\n--- Validating XML against X3D 4.0 Schema (using libxml2) ---")
	err = validateXMLWithSchema(output, schemaFilename)
	if err != nil {
		fmt.Printf("--- Invalid Generated XML ---\n%s\n---------------------------\n", string(output))
		log.Fatalf("Schema validation failed for generated XML: %v", err)
	}
	fmt.Println("✅ XML is valid against the X3D 4.0 schema!")
	*/
	filename := "../data/geobubbles.new.go.x3d"
	file, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		return
	}
	defer file.Close() // Ensure the file is closed when the function exits

	// Write the string content to the file
	header := "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE X3D PUBLIC \"ISO//Web3D//DTD X3D 4.0//EN\" \"https://www.web3d.org/specifications/x3d-4.0.dtd\">\n"
	_, err = file.WriteString(header)
	if err != nil {
		fmt.Printf("Error writing header to file: %v\n", err)
		return
	}

	_, err = file.Write(output)
	if err != nil {
		fmt.Printf("Error writing bytes to file: %v\n", err)
		return
	}

	fmt.Printf("\n✅ Scene successfully written to %s\n", filename)
	fmt.Printf("   You can view the generated XML below:\n\n%s\n", string(output))
}